### PR TITLE
[LEC] Executor cache: keep block speculation result inside executor

### PR DESCRIPTION
**Pull request by lightmark**
## Motivation

This PR made a huge refactoring to `Executor` to cache speculation result inside executor instead of returning back to consensus cached inside the block_tree maintained by consensus, required by #2519 as the first milestone.

The reason for this refactoring was explained in #2519 . In short, some speculation result is unserializable so can not be sent over wire. Since `Executor` will be a standalone service in a container, all the current API will become RPC API called via messages over wire. The only option for us is to keep unserializable speculation results inside.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

CI

### CODE REVIEW

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 1]
```diff
@@ -85,34 +87,41 @@ impl<T: Payload> BlockStore<T> {
 
     fn build_block_tree(
         root: RootInfo<T>,
-        root_executed_trees: ExecutedTrees,
+        root_data: RootMetaData,
```
Comment (davidiw) [inline consensus/src/chained_bft/block_storage/block_store.rs]: Metadata

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 2]
```diff
@@ -265,21 +270,23 @@ impl<T: Payload> BlockStore<T> {
...
-                parent_block.output().executed_trees().clone(),
-                parent_block.output().validators().clone(),
+        let state_compute_result = if parent_block.compute_result().has_reconfiguration() {
+            StateComputeResult::new(
+                parent_block.compute_result().root_hash(),
+                parent_block.compute_result().frozen_subtree_roots().clone(),
+                parent_block.compute_result().num_leaves(),
+                parent_block.compute_result().validators().clone(),
+                vec![], /* compute_status */
+                vec![], /* transaction_info_hashes */
```
Comment (davidiw) [inline consensus/src/chained_bft/block_storage/block_store.rs]: would it make sense to make this a tuple via say a zip operation or something?

File: consensus/src/chained_bft/persistent_liveness_storage.rs [hunk 3]
```diff
@@ -147,14 +149,39 @@ impl LedgerRecoveryData {
     }
 }
 
+pub struct RootMetaData {
```
Comment (davidiw) [inline consensus/src/chained_bft/persistent_liveness_storage.rs]: Metadata

Comment (davidiw) [reply consensus/src/chained_bft/persistent_liveness_storage.rs]: This may very well end up in consensus types, since we'll need to keep track of the latest root hash there too...

File: consensus/src/chained_bft/test_utils/mod.rs [hunk 4]
```diff
@@ -221,7 +221,7 @@ pub fn consensus_runtime() -> runtime::Runtime {
         ::libra_logger::Logger::new().level(Level::Debug).init();
     }
     // setup timeout for tests
-    crash_handler::setup_panic_handler();
+    // crash_handler::setup_panic_handler();
```
Comment (davidiw) [inline consensus/src/chained_bft/test_utils/mod.rs]: sup?

File: consensus/src/consensus_provider.rs [hunk 5]
```diff
@@ -39,7 +39,7 @@ pub fn make_consensus_provider(
     node_config: &mut NodeConfig,
     network_sender: ConsensusNetworkSender<Vec<SignedTransaction>>,
     network_receiver: ConsensusNetworkEvents<Vec<SignedTransaction>>,
-    executor: Arc<Executor<LibraVM>>,
+    executor: Arc<Mutex<Executor<LibraVM>>>,
```
Comment (davidiw) [inline consensus/src/consensus_provider.rs]: oh gosh no....

Comment (davidiw) [reply consensus/src/consensus_provider.rs]: but it makes sense...

File: execution/executor-types/src/lib.rs [hunk 6]
```diff
@@ -44,31 +44,37 @@ pub struct StateComputeResult {
     /// for StateMachineReplication, which is merely passing it between StateComputer and
     /// TxnManager.
     compute_status: Vec<TransactionStatus>,
+    /// The transaction info hashes of all success txns.
+    transaction_info_hashes: Vec<HashValue>,
 }
 
 impl StateComputeResult {
     pub fn new(
-        accu_root_hash: HashValue,
```
Comment (davidiw) [inline execution/executor-types/src/lib.rs]: why you make this type of critical change here? this is part of the refactoring you could have done in another PR :P

File: types/src/block_info.rs [hunk 7]
```diff
@@ -155,10 +154,11 @@ impl Display for BlockInfo {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
         write!(
             f,
-            "BlockInfo: [epoch: {}, round: {}, id: {}, version: {}, timestamp (us): {}, next_validator_set: {}]",
+            "BlockInfo: [epoch: {}, round: {}, id: {}, executed_state_id: {}, version: {}, timestamp (us): {}, next_validator_set: {}]",
```
Comment (davidiw) [inline types/src/block_info.rs]: why aren't we calling this root_hash?

File: secure/key-manager/src/tests.rs [hunk 8]
```diff
@@ -163,34 +162,38 @@ impl LibraInterface for TestLibraInterface {
...
     // the clock is always refreshed for this set of transactions.
     let block_metadata = BlockMetadata::new(block_id, 0, clock, vec![], node.account);
     let prologue = Transaction::BlockMetadata(block_metadata);
     block.insert(0, prologue);
 
-    let output = node
+    let result = node
         .executor
-        .execute_block(block_id, block.clone(), &committed_trees, &committed_trees)
+        .execute_block((block_id, block), node.executor.committed_block_id())
```
Comment (davidiw) [inline secure/key-manager/src/tests.rs]: could we not have made this into another PR?

File: execution/executor/src/lib.rs [hunk 9]
```diff
@@ -40,29 +42,227 @@ use std::{
...
+    sync::{Arc, Mutex, Weak},
 };
 use storage_client::{StorageRead, StorageWrite, VerifiedStateView};
+use storage_proto::StartupInfo;
 use tokio::runtime::Runtime;
 
 static OP_COUNTERS: Lazy<libra_metrics::OpMetrics> =
     Lazy::new(|| libra_metrics::OpMetrics::new_and_registered("executor"));
 
+struct SpeculationBlock {
```
Comment (davidiw) [inline execution/executor/src/lib.rs]: I thought the plan was to wrap the SpeculativeBlock in a container outside the core Executor that way we didn't have to modify state sync. I'm not clear why state sync needs speculation at all though... since I thought it would be doing execute and commit?

Comment (davidiw) [reply execution/executor/src/lib.rs]: This really should be in a separate module not adding 350 lines to your lib

File: execution/executor/src/lib.rs [hunk 10]
```diff
@@ -40,29 +42,227 @@ use std::{
...
+        *self.block_map.lock().unwrap() = HashMap::new();
+    }
+
+    fn add_block(
+        &mut self,
+        parent_block_id: HashValue,
+        block: (
+            HashValue, /* block_id */
+            Vec<Transaction>,
+            ProcessedVMOutput,
```
Comment (davidiw) [inline execution/executor/src/lib.rs]: let's make this a type?

File: execution/executor/src/lib.rs [hunk 11]
```diff
@@ -40,29 +42,227 @@ use std::{
...
+            self.get_block(&parent_block_id)?
+                .lock()
+                .unwrap()
+                .add_child(block);
+        }
+        Ok(())
+    }
+
+    fn prune(&mut self, committed_ledger_info: &LedgerInfo) -> Result<()> {
+        let arc_latest_committed_block =
```
Comment (davidiw) [inline execution/executor/src/lib.rs]: maybe instead we sort these by version and we eliminate all versions earlier than the newly committed version? alternatively it isn't really clear to me why we're cloning if we're pruning.

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 2 above)
Comment (lightmark) [reply consensus/src/chained_bft/block_storage/block_store.rs]: They are not always of the same length, dude

File: consensus/src/consensus_provider.rs (hunk 5 above)
Comment (lightmark) [reply consensus/src/consensus_provider.rs]: I'll leave this for later change.

File: execution/executor-types/src/lib.rs (hunk 6 above)
Comment (lightmark) [reply execution/executor-types/src/lib.rs]: Yes, but this one seems trivial. The separation of it doesn't help a lot.

File: types/src/block_info.rs (hunk 7 above)
Comment (lightmark) [reply types/src/block_info.rs]: That's how consensus call it. cc: @zekun000

File: secure/key-manager/src/tests.rs (hunk 8 above)
Comment (lightmark) [reply secure/key-manager/src/tests.rs]: how...
I think if we made this into another PR, this PR will not pass CI. This test will fail since it calls executor's api.

File: execution/executor/src/lib.rs (hunk 9 above)
Comment (lightmark) [reply execution/executor/src/lib.rs]: StateSync has nothing to do with `SpeculationBlock`. The plan was to split Executor into two and the one used by state sync will not have this block tree. But I agree the idea of putting it into a standalone module.

File: execution/executor/src/lib.rs (hunk 10 above)
Comment (lightmark) [reply execution/executor/src/lib.rs]: This is used internally with only one callsite. IMO I don't think it is quite necessary and not suitable for this PR :)

File: execution/executor/src/lib.rs (hunk 11 above)
Comment (lightmark) [reply execution/executor/src/lib.rs]: I didn't get your question. I use Weak pointer in the tree so after update the `heads`, reference counting of all to-be-pruned branches will be cleaned automatically by reference counting.

File: execution/executor/src/lib.rs [hunk 12]
```diff
@@ -51,18 +54,21 @@ static OP_COUNTERS: Lazy<libra_metrics::OpMetrics> =
...
     storage_read_client: Arc<dyn StorageRead>,
     storage_write_client: Arc<dyn StorageWrite>,
-
+    cache: SpeculationCache,
     phantom: PhantomData<V>,
 }
 
 impl<V> Executor<V>
 where
     V: VMExecutor,
```
Comment (davidiw) [inline execution/executor/src/lib.rs]: Do we have other VMExecutor? Yes, we do... Is it useful?

File: execution/executor/src/lib.rs [hunk 13]
```diff
@@ -110,50 +119,59 @@ where
...
-        let pre_genesis_trees = ExecutedTrees::new_empty();
-        let output = self
+        let result = self
             .execute_block(
-                *GENESIS_BLOCK_ID,
-                genesis_txns.clone(),
-                &pre_genesis_trees,
-                &pre_genesis_trees,
+                (
+                    HashValue::zero(), /* match with the id in BlockInfo::genesis(...) */
```
Comment (davidiw) [inline execution/executor/src/lib.rs]: why not just us //

File: execution/executor/src/speculation_cache/mod.rs [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+        Self {
+            synced_trees: ExecutedTrees::new_empty(),
+            committed_trees: ExecutedTrees::new_empty(),
+            heads: vec![],
+            block_map: Arc::new(Mutex::new(HashMap::new())),
+            committed_block_id: *PRE_GENESIS_BLOCK_ID,
+        }
+    }
+
+    pub(crate) fn new_with_startup_info(startup_info: StartupInfo) -> Self {
```
Comment (davidiw) [inline execution/executor/src/speculation_cache/mod.rs]: these pub(crate) are meaningless if you define you struct as pub(crate)

File: execution/executor/src/speculation_cache/test.rs [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    let num_blocks = cache.block_map.lock().unwrap().len();
+    assert_eq!(num_blocks, 4);
+    assert_ne!(cache.committed_block_id, id(1));
+}
+
+#[test]
+fn test_add_duplicate_block() {
+    let mut cache = create_cache();
+    assert!(cache.add_block(id(1), gen_block(id(7))).is_ok());
+    assert!(cache.add_block(id(1), gen_block(id(7))).is_ok());
```
Comment (davidiw) [inline execution/executor/src/speculation_cache/test.rs]: if you use unwrap here instead... you'll get the error message upon failure.

File: execution/executor/src/speculation_cache/mod.rs (hunk 14 above)
Comment (lightmark) [reply execution/executor/src/speculation_cache/mod.rs]: I am not sure what's the convention here.

File: execution/executor/src/lib.rs (hunk 12 above)
Comment (lightmark) [reply execution/executor/src/lib.rs]: There is a mock one.

File: consensus/src/chained_bft/block_storage/block_store.rs [hunk 16]
```diff
@@ -85,34 +87,41 @@ impl<T: Payload> BlockStore<T> {
...
+            root_metadata.version(),
         );
         assert_eq!(
             root_qc.certified_block().executed_state_id(),
-            root_executed_trees.state_id(),
+            root_metadata.accu_hash,
             "root qc state id {} doesn't match committed trees {}",
             root_qc.certified_block().executed_state_id(),
-            root_executed_trees.state_id(),
+            root_metadata.accu_hash,
```
[the remark this answers is not in the document]
Comment (lightmark) [reply consensus/src/chained_bft/block_storage/block_store.rs]: Yes, it is a tricky thing we need to agree on across the whole repo. cc: @zekun000

**Comment by lightmark** [comment]
> I guess you get an error on commit if storage and cache are out of sync and this is the same behavior consensus would have exhibited? We don't currently have a recovery process here, do we?

No. That is out of the scope of this PR.

File: consensus/src/chained_bft/block_storage/block_store.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/chained_bft/block_storage/block_store.rs]: propose something?
